apturl.8
.TH APTURL 8 "September 01, 2008"

.SH NAME
apturl \- graphical apt\-protocol interpreting package installer

.SH SYNOPSIS
\fBapturl\fP [\fBoptions...\fP] <\fBURL\fP>

.SH DESCRIPTION
\fBapturl\fP is a simple graphical application that takes an URL (following the
apt\-protocol) as a command line option, parses it and carries out the
operations that the URL describes (that is, it asks the user if he wants the
indicated packages to be installed and if the answer is positive does so).

.SH USAGE
\fBapturl\fP just needs an URL conforming with the apt-protocol in order to
work. Additionally, it recognizes the following options:
.TP
\fB\-p, \-\-http\-proxy\fP
Use the given HTTP proxy in order to download the packages.

.SH EXAMPLES
.TP
apturl apt:pidgin,pidgin-plugin-pack
Installs Pidgin and Pidgin Plugin Pack (if the user confirms).
.TP
apturl apt:freevial?section=universe
Enables the "universe" component and installs package Freevial.
.TP
apturl apt:adobe-flashplugin?channel=lucid-partner
Enables the "partner" repository and installs package adobe-flashplugin.
Available repositories are listed in /usr/share/app-install/channels/.
.TP
apturl apt+http://launchpad.net/~mvo/ppa?package=2vcard
Installs 2vcard from the indicated PPA (if the user confirms), and afterward
asks if the PPA should be removed again or it should remain enabled.
.br
Warning: This is currently disabled because of security concerns.

.SH EXIT CODES
.TP
0
All requested packages have been successfully installed or were already
installed.
.TP
1
The user aborted the installation of one or more packages.
.TP
2
There was an error doing the requested operations, or the user aborted the
installation while Synaptic (and not apturl itself) was working.
.TP
3
The arguments with which \fBapturl\fP was called were incorrect (e.g., it
was called without any argument or the given URL was wrong).

.SH BUGS
Please report any bug on https://bugs.launchpad.net/ubuntu/+source/apturl.

.SH SEE ALSO
\fBsynaptic\fR

.SH AUTHORS
\fBapturl\fP was written by Michael Vogt <michael.vogt@ubuntu.com> for
Canonical, Ltd. and this manual page by Siegfried-A. Gevatter
<rainct@ubuntu.com>.
.PP
It is released under the GNU General Public License, version 2 or any
later version published by the Free Software Foundation.


debian/changelog
apturl (0.5.2+linuxmint3) sarah; urgency=medium

  * Updated translations

 -- Clement Lefebvre <root@linuxmint.com>  Thu, 05 May 2016 18:42:53 +0100

apturl (0.5.2+linuxmint2) sarah; urgency=medium

  * Remove channel/section support and dependencies on Ubuntu components and webkit

 -- Clement Lefebvre <root@linuxmint.com>  Thu, 05 May 2016 16:54:28 +0100

apturl (0.5.2+linuxmint1) sarah; urgency=medium

  * Removed QT interface
  * Remove backends (will be reimplemented without dependencies on Ubuntu components)
  * Specify Gtk version in gi
  * Implemented synaptics installations

 -- Clement Lefebvre <root@linuxmint.com>  Thu, 05 May 2016 16:32:05 +0100

apturl (0.5.2ubuntu11) xenial; urgency=medium

  * Drop webkit1 support, we're switching to pure webkit2 now

 -- Iain Lane <iain@orangesquash.org.uk>  Wed, 13 Jan 2016 11:48:22 +0000

apturl (0.5.2ubuntu10) xenial; urgency=medium

  * debian/apturl.install:
    - don't install the gconf schemas, it's probably useful to no one
      nowadays (would only work for applications using the GNOME2 libgnome)
      and creates a depends on gconf which we want to drop from the iso

 -- Sebastien Bacher <seb128@ubuntu.com>  Wed, 16 Dec 2015 11:23:10 +0100

apturl (0.5.2ubuntu9) wily; urgency=medium

  * data/apturl-gtk.ui: set the dimension on the scrolledwindow rather
    than the textview so the description is showing with a non zero height 

 -- Sebastien Bacher <seb128@ubuntu.com>  Thu, 15 Oct 2015 17:56:46 +0100

apturl (0.5.2ubuntu8) wily; urgency=medium

  * Use webkit2 in addition to webkit1. (LP: #1469221)

 -- Iain Lane <iain@orangesquash.org.uk>  Tue, 30 Jun 2015 12:59:59 +0100

apturl (0.5.2ubuntu7) wily; urgency=medium

  * Don't include Debian version in the python module version (LP: #1465549)

 -- Zygmunt Krynicki <zygmunt.krynicki@canonical.com>  Fri, 26 Jun 2015 11:01:58 +0200

apturl (0.5.2ubuntu6) vivid; urgency=medium

  [ Tim Lunn ]
  * Don't use legacy icon names, these are no longer available in
    adwaita-icon-theme. (LP: #1437150) 

  [ Iain Lane ]
  * For dialog-question, go back to using the stock icon name as Humanity
    doesn't yet have an icon with the new name.

 -- Tim Lunn <tim@feathertop.org>  Thu, 02 Apr 2015 13:53:05 +0100

apturl (0.5.2ubuntu5) vivid; urgency=medium

  * Switch gnome-icon-theme to adwaita-icon-theme, which is its
    successor. (LP: #1414613)

 -- Iain Lane <iain@orangesquash.org.uk>  Mon, 26 Jan 2015 14:55:39 +0000

apturl (0.5.2ubuntu4) trusty; urgency=medium

  * Drop libgtk2-perl Recommends to Suggests.

 -- Dimitri John Ledkov <xnox@ubuntu.com>  Sun, 16 Mar 2014 02:56:58 +0000

apturl (0.5.2ubuntu3) trusty; urgency=low

  * AptUrl/AptUrl.py: don't use decode on a str (lp: #1050826)

 -- Sebastien Bacher <seb128@ubuntu.com>  Wed, 11 Dec 2013 11:18:25 +0100

apturl (0.5.2ubuntu2) saucy; urgency=low

  * adjust backend to reuse the bits from update-manager (LP: #1200775)

 -- Michael Vogt <michael.vogt@ubuntu.com>  Tue, 08 Oct 2013 20:14:22 +0200

apturl (0.5.2ubuntu1) raring; urgency=low

  * AptUrl/gtk/backend/__init__.py: Drop "action-done" signal definition;
    update-manager's GTK backend already defines it with different arguments,
    so this is ineffective and confusing.
  * AptUrl/gtk/GtkUI.py, _on_backend_done(): Add additional error signal
    arguments, as defined by update-manager's backend.
  * AptUrl/gtk/backend/InstallBackend{Aptdaemon,Synaptic}.py: Call action-done
    signal with error arguments. (LP: #1103024)
  * data/apturl-gtk.ui: Drop obsolete "has_separator" property. Fixes a
    warning.

 -- Martin Pitt <martin.pitt@ubuntu.com>  Tue, 09 Apr 2013 13:23:37 +0200

apturl (0.5.1ubuntu6) quantal; urgency=low

  * AptUrl/gtk/backend/InstallBackendAptdaemon.py: Work around
    incompatible Update Manager API change (as described in Launchpad
    bug #1026257).  LP: #1020980

 -- Barry Warsaw <barry@ubuntu.com>  Wed, 18 Jul 2012 13:38:02 -0400

apturl (0.5.1ubuntu5) quantal; urgency=low

  * debian/control: Move python3-aptdaemon dependency to apturl binary
    package (from apturl-common) since only the gtk backend needs it.

 -- Barry Warsaw <barry@ubuntu.com>  Fri, 15 Jun 2012 09:49:13 -0400

apturl (0.5.1ubuntu4) quantal; urgency=low

  [ Brian Murray ]
  * Port to Python 3:
    - Use Python 3 style print functions
  * AptUrl/Parser.py:
    - ensure MAX_URL_LEN/10 returns an int

  [ Colin Watson ]
  * AptUrl/gtk/backend/InstallBackendSynaptic.py:
    - Keep a reference to the data tuple passed to GObject.child_watch_add
      to avoid attempts to destroy it without a thread context.
    - Open temporary synaptic selections file in text mode.

  [ Barry Warsaw ]
  * Additional Python 3 fixes:
    + Port setup.py to Python 3.
    + Use the new python-apt API since the legacy API is not available in
      Python 3.
    + ki18n() takes bytes.
    + Fix relative imports so the code can be run from source.
    + Fixed some additional packaging paths.
    + Change #! lines to use python3.
    + Changed debian/control and debian/rules to use python3.
    + Change debian/compat == 9
    + apturl now must require python3-aptdaemon.gtk3widgets (i.e. not |
      synaptic) due to the Python 3 port of update-manager.
  * Other changes:
    + Remove the need for threads in the KDE front-end.
    + Enable running the test suite via `python3 -m unittest discover`
    + i18n updates.
    + Various and sundry pyflakes, whitespace, style, line length, and
      spelling fixes.

 -- Brian Murray <brian@ubuntu.com>  Mon, 11 Jun 2012 10:32:52 -0700

apturl (0.5.1ubuntu3) precise; urgency=low

  * Take the unicode representation of the parser error message instead
    of the string representation, so that we get clean i18n handling.
    LP: #911144.

 -- Steve Langasek <steve.langasek@ubuntu.com>  Sat, 24 Mar 2012 02:36:28 -0700

apturl (0.5.1ubuntu2) precise; urgency=low

  * debian/control: Fix dependencies: python-gobject → python-gi.


AptUrl/AptUrl.py
# Copyright (c) 2007-2008 Canonical
#
# AUTHOR:
# Michael Vogt <mvo@ubuntu.com>
# With contributions by Siegfried-A. Gevatter <rainct@ubuntu.com>
#
# This file is part of AptUrl
#
# AptUrl is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as published
# by the Free Software Foundation; either version 2 of the License, or (at
# your option) any later version.
#
# AptUrl is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with AptUrl; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

import apt
import apt_pkg

from . import Parser
from . import Helpers
from .Helpers import _

from optparse import OptionParser

import os
import os.path

# adding new repositories is currently disabled because some people have
# security concerns about this feature
allow_new_repositories = False

# return codes
(RESULT_OK,
 RESULT_CANCELT,
 RESULT_ERROR,
 RESULT_BADARGS) = list(range(4))

class AptUrlController(object):

    def __init__(self, ui):
        self.ui = ui

    def openCache(self):
        try:
            self.cache = apt.Cache()
        except SystemError as strerr:
            if not '/etc/apt/sources.list' in str(strerr):
                raise
            self.ui.error(_("Invalid /etc/apt/sources.list file"), strerr)
            return False
        if self.cache._depcache.broken_count > 0:
            err_header = _("Software index is broken")
            err_body = _("This is a major failure of your software "
                         "management system. Please check for broken packages "
                         "with synaptic, check the file permissions and "
                         "correctness of the file '/etc/apt/sources.list' and "
                         "reload the software information with: "
                         "'sudo apt-get update' and 'sudo apt-get install -f'."
                         )
            self.ui.error(err_header, err_body)
            return False
        return True

    def parseArgs(self):
        parser = OptionParser()
        parser.add_option("-p", "--http-proxy", dest="http_proxy",
                          default=None, help="use http proxy")
        (options, args) = parser.parse_args()

        # eval and add proxy
        if options.http_proxy is not None:
            proxy = options.http_proxy
            if not ":" in proxy:
                proxy += ":3128"
            os.environ["http_proxy"] = "http://%s" % proxy

        # parse
        try:
            apturl_list = Parser.parse(args[0])
        except IndexError as e:
            self.ui.error(_("Need a url to continue, exiting"))
            return []
        except Parser.InvalidUrlException as e:
            self.ui.error(_("Invalid url: '%s' given, exiting") % e.url,
                          str(e))
            return []
        return (apturl_list)

    def verifyInstall(self, apturl):
        " verify that the install package actually is installed "
        # check if the package got actually installed
        self.openCache()
        pkg = self.cache[apturl.package]
        if (not pkg.is_installed or
            pkg._pkg.current_state != apt_pkg.CURSTATE_INSTALLED or
            self.cache._depcache.broken_count > 0):
            return False
        return True

    def main(self):
        # global return code
        ret = RESULT_OK
        ui = self.ui

        # parse arguments
        apturl_list = self.parseArgs()
        if not apturl_list:
            return RESULT_BADARGS

        # open cache
        if not self.openCache():
            return RESULT_ERROR

        # now go over the url list
        for apturl in apturl_list:
            # FIXME: move this code block into a func like
            #        evalAptUrl()

            if not apturl.schema in ("apt", "apt+http"):
                self.ui.error(_("Can not deal with protocol '%s' ")
                              % apturl.schema)
                continue

            if apturl.refresh is not None:
                ui.doUpdate()
                if not self.openCache():
                    return RESULT_ERROR

            # now check the package
            if apturl.package not in self.cache:
                try:
                    package_in_cache = bool(self.cache._cache[apturl.package])
                except KeyError:
                    package_in_cache = False
                if package_in_cache:
                    ui.error(_("Package '%s' is virtual.") % apturl.package)
                    continue
                else:
                    ui.error(_("Could not find package '%s'.")
                             % apturl.package)
                    continue

            if (self.cache[apturl.package].is_installed and
                apturl.minver is None):
                ui.message(_("Package '%s' is already installed")
                           % apturl.package)
                continue

            # ask the user
            pkg = self.cache[apturl.package]
            (sum, desc, homepage) = Helpers.parse_pkg(pkg)
            if not ui.askInstallPackage(apturl.package, sum, desc, homepage):
                ret = RESULT_CANCELT
                continue

            # try to install it
            try:
                self.cache[apturl.package].mark_install()
            except SystemError as e:
                ui.error(_("Can not install '%s' (%s) ") % (apturl.package, e))
                continue
            if apturl.minver is not None:
                verStr = self.cache[apturl.package].candidate.version
                if apt_pkg.version_compare(verStr, apturl.minver) < 1:
                    ui.error(_(
                        "Package '%s' requests minimal version '%s', but "
                        "only '%s' is available") % (apturl.package,
                                                     apturl.minver,
                                                     verStr))
                    continue

            # install it
            ui.doInstall(apturl)

            if not self.verifyInstall(apturl):
                ret = RESULT_ERROR

        # return values
        return ret
